src/accounts_refactor.rs

```diff
@@ -0,0 +1,339 @@
+    type PubKey: Clone + DeserializeOwned + Hash + Eq + Serialize;
+    fn pub_key(&self) -> Self::PubKey;
+}
+
+impl KeyPair for ViewerKeyPair {
+    type PubKey = ViewerPubKey;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.pub_key()
+    }
+}
+
+impl KeyPair for FreezerKeyPair {
+    type PubKey = FreezerPubKey;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.pub_key()
+    }
+}
+
+impl KeyPair for UserKeyPair {
+    // The PubKey here is supposed to be a conceptual "primary key" for looking up UserKeyPairs. We
+    // typically want to look up UserKeyPairs by Address, not PubKey, because if we have a PubKey we
+    // can always get an Address to do the lookup.
+    type PubKey = UserAddress;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.address()
+    }
+}
+
+#[derive(Clone, Deserialize, Serialize)]
+#[serde(bound = "Key: DeserializeOwned + Serialize")]
+pub struct Account<L: Ledger, Key: KeyPair> {
+    /// The account key.
+    key: Key,
+    /// The account description.
+    pub description: String,
+    /// Whether the account is used.
+    pub used: bool,
+    /// Optional ledger scan.
+    scan: Option<BackgroundKeyScan<L>>,
+    /// The list of asset codes of the account.
+    assets: Vec<AssetCode>,
+    /// The list of records of the account.
+    records: Vec<RecordInfo>,
+    /// The table of balances with corresponding asset code.
+    balances: HashMap<AssetCode, U256>,
+    /// The time when the account was created.
+    created_time: DateTime<Local>,
+    /// The last time when the account was modified.
+    modified_time: DateTime<Local>,
+}
+
+impl<L: Ledger, Key: KeyPair> Account<L, Key> {
+    #![allow(dead_code)]
+
+    /// Get the account key.
+    pub fn key(&self) -> &Key {
+        &self.key
+    }
+
+    /// Get the account public key.
+    pub fn pub_key(&self) -> Key::PubKey {
+        self.key.pub_key()
+    }
+
+    /// Get the account description.
+    pub fn description(&self) -> String {
+        self.description.clone()
+    }
+
+    /// Check whether the account is used.
+    pub fn used(&self) -> bool {
+        self.used
+    }
+
+    /// Get the optional ledger scan.
+    pub fn scan(&self) -> Option<BackgroundKeyScan<L>> {
+        self.scan.clone()
+    }
+
+    /// Get the list of asset codes.
+    pub fn assets(&self) -> Vec<AssetCode> {
+        self.assets.clone()
+    }
+
+    /// Get the list of records.
+    pub fn records(&self) -> Vec<RecordInfo> {
+        self.records.clone()
+    }
+
+    /// Get the table of balances.
+    pub fn balances(&self) -> HashMap<AssetCode, U256> {
+        self.balances.clone()
+    }
+
+    /// Get the created time.
+    pub fn created_time(&self) -> DateTime<Local> {
+        self.created_time
+    }
+
+    /// Get the modified time.
+    pub fn modified_time(&self) -> DateTime<Local> {
+        self.modified_time
+    }
+}
+
+pub type AccountsStore<L, Key, PubKey> = KeyValueStore<PubKey, Account<L, Key>>;
```

Comment:
We could make this `pub type AccountsStore<L, Key> = KeyValueStore<Key::PubKey, Account<L, Key>>` to get rid of the dependent parameter

Reply:
I tried this but got an error that `Key::PubKey` wasn't defined. I resolved the "undefined" error by `AccountsStore<L, Key: KeyPair>`, but there was a clippy warning that the `Key: KeyPair` would be ignored since `AccountStore` was a type alias.

Reply:
Huh. I thought it would work as long as it was constrained by `Key: KeyPair` wherever we instantiate the type alias. I guess never mind then. Definitely not a big deal